lib/galaxy/util/template.py

```diff
@@ -15,4 +49,60 @@ def fill_template(template_text, context=None, **kwargs):
         raise TypeError("Template text specified as None to fill_template.")
     if not context:
         context = kwargs
-    return unicodify(Template(source=template_text, searchList=[context]))
+    klass = Template.compile(source=template_text, compilerClass=compiler_class)
+    t = klass(searchList=[context])
+    try:
+        return unicodify(t)
+    except NotFound as e:
+        if first_exception is None:
+            first_exception = e
+        if retry > 0 and sys.version_info.major > 2:
+            tb = e.__traceback__
+            last_stack = traceback.extract_tb(tb)[-1]
+            if last_stack.name == '<listcomp>':
+                # On python 3 list,dict and set comprehensions as well as generator expressions
+                # have their own local scope, which prevents accessing frame variables in cheetah.
+                # We can work around this by replacing `$var` with `var`
+                var_not_found = e.args[0].split("'")[1]
+                replace_str = 'VFFSL(SL,"%s",True)' % var_not_found
+                lineno = last_stack.lineno - 1
+                module_code = t._CHEETAH_generatedModuleCode.splitlines()
+                module_code[lineno] = module_code[lineno].replace(replace_str, var_not_found)
+                module_code = "\n".join(module_code)
+                compiler_class = create_compiler_class(module_code)
+                return fill_template(template_text=template_text,
+                                     context=context,
+                                     retry=retry - 1,
+                                     compiler_class=compiler_class,
+                                     first_exception=first_exception
+                                     )
+        raise first_exception or e
+    except Exception as e:
+        if first_exception is None:
+            first_exception = e
+        if not futurized:
```

Comment:
```suggestion
        if not futurized and sys.version_info.major > 2:
```

Reply:
Thanks, I've added that to https://github.com/galaxyproject/galaxy/pull/7867/commits/8f692cbebd14bf4108ed050867aeb60bc54425d5